c/tskit/tables.c

```diff
@@ -5341,6 +5334,547 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+            // Create a SegmentList holding all of the sample segments descending from
+            // seg.
+            u = seg->node;
+            if (self->is_sample[u]) {
+                ret = ibd_finder_enqueue_segment(self, seg->left, seg->right, seg->node);
+                if (ret != 0) {
+                    goto out;
+                }
+            } else {
+                if (self->ancestor_map_head[u] != NULL) {
+                    s = self->ancestor_map_head[u];
+                    while (s != NULL) {
+                        if (seg->left > s->left) {
+                            intvl_l = seg->left;
+                        } else {
+                            intvl_l = s->left;
+                        }
+                        if (seg->right < s->right) {
+                            intvl_r = seg->right;
+                        } else {
+                            intvl_r = s->right;
+                        }
+                        // Add to the segment queue.
+                        if (intvl_r - intvl_l > 0) {
+                            ret = ibd_finder_enqueue_segment(
+                                self, intvl_l, intvl_r, s->node);
+                            if (ret != 0) {
+                                goto out;
+                            }
+                        }
+                        s = s->next;
+                    }
+                }
+            }
+
+            // Calculate new ibd segments descending from the current parent.
+            if (self->segment_queue_size > 0) {
+                ret = tsk_ibd_finder_calculate_ibd(self, current_parent);
+            }
+            if (ret != 0) {
+                goto out;
+            }
+
+            // For samples that appear in the parent column of the edge table
+            if (self->is_sample[current_parent] && parent_should_be_added) {
+                ret = ibd_finder_enqueue_segment(
+                    self, 0, self->sequence_length, current_parent);
+                if (ret != 0) {
+                    goto out;
+                }
+                ret = tsk_ibd_finder_calculate_ibd(self, current_parent);
+                if (ret != 0) {
+                    goto out;
+                }
+                parent_should_be_added = false;
+            }
+
+            // if (j != num_edges - 1 && input_edges->parent[j + 1] != current_parent) {
+            //     parent_should_be_added = true;
+            //     for (k = 0; k < self->num_nodes; k++) {
+            //         // This feels rly inefficient now... a better way?
+            //         if (current_parent == self->oldest_parent[k]) {
+            //             // The above causes a segfault at present.
+            //             printf("TODO: Memory can be freed here, think about
+            //             later...\n");
+            //         }
+            //     }
+            // }
+        }
+    }
+
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
```

Comment:
I don't think this should be part of the external API. Mark it as static and put it further up in the file (static functions must be defined before they are used)